[libc++] std::atomic primary template should not have a `difference_type` member type

dalg24 opened the pull request:
The test would not check its absence and the code path intended for pointer was never actually instantiated.
I added a few pointer types since there was no coverage.

dalg24 committed:
Fix tests for std::atomic member types

dalg24 committed:
Fix std::atomic primary template should not have difference_type

dalg24 commented on libcxx/test/std/atomics/types.pass.cpp:
```diff
@@ -29,163 +29,173 @@
 };
 
 template <class A>
-struct test_atomic<A*, false>
```
This would never occur since we were instantiating with `A = std::atomic<T>`

dalg24 commented:
FWIW it looks like it was only broken in libc++ https://godbolt.org/z/x8G5qqvzW

philnik777 commented on libcxx/test/std/atomics/types.pass.cpp:
```diff
@@ -29,163 +29,173 @@
 #  include <thread>
 #endif
 
```
Could you format the test in a separate commit and revert the changes here? It's really hard to see what actually changed.

dalg24 replied:
Will do.
I realized I forgot to run in `libcxx`, I had just ran `std`...

dalg24 replied:
Done

dalg24 replied:
It is also somewhat readable if you hide whitespaces

philnik777 reviewed (commented):
The implementation changes look good, but I haven't looked closely at the test yet.

dalg24 committed:
Update expected error for atomic_fetch_{add,sub}[_explicit] with pointers to member function

dalg24 committed:
Apply clang-format

dalg24 committed:
Add test with function pointer

dalg24 committed:
Fixup for cxx03

Work around `error: unknown type name 'constexpr'`.
Not sure that is the right approach since `<atomic>` is C++11...

dalg24 commented:
> The implementation changes look good, but I haven't looked closely at the test yet.

Would you give me some directions on how to approach the cxx03 failures?
`<atomic>` is C++11.  Should I just add `// UNSUPPORTED: c++03` directives where I got failures?

ldionne commented:
> > The implementation changes look good, but I haven't looked closely at the test yet.
> 
> I could use directions on how to approach the cxx03 failures? `<atomic>` is C++11. Should I just add `// UNSUPPORTED: c++03` directives where I got failures?

For that test you should add `// XFAIL: FROZEN-CXX03-HEADERS-FIXME`

dalg24 commented on libcxx/test/std/atomics/types.pass.cpp:
```diff
@@ -29,163 +29,176 @@
+template <class A, bool IntegralOrFloating, bool Pointer>
+struct test_atomic {
+  test_atomic() {
+    static_assert(!IntegralOrFloating || !Pointer, "");
```
No, I meant it, that is `!(IntegralOrFloating && Pointer)`, as a sanity check that (true, true) never happens.
But with your question I see that this is confusing.
Would you like me to drop it?

dalg24 replied:
I applied your suggestion

dalg24 commented on libcxx/test/std/atomics/types.pass.cpp:
```diff
@@ -29,163 +29,176 @@
+template <typename T>
+struct has_difference_type<T, myvoid_t<typename T::difference_type> > : std::true_type {};
+
+template <class A, bool IntegralOrFloating, bool Pointer>
```
Totally agree with you, was refraining to change things in that test.
Will apply once I get to a place with clang-format available.

dalg24 replied:
Done

dalg24 committed:
Apply suggested changes from review

dalg24 committed:
Add XFAIL directive for the frozen cxx03 tests per review

dalg24 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into atomic_difference_type

dalg24 committed:
Fixup missing typename keyword

dalg24 committed:
Fixup std::atomic<Floating>::difference_type was introduced in C++20

Drop guards for C++ version greater equal 17 to increase coverage.

ldionne committed:
Merge branch 'main' into atomic_difference_type

ldionne committed:
XFAIL test in C++03 frozen

dalg24 commented on libcxx/test/std/atomics/types.pass.cpp:
```diff
@@ -17,6 +17,9 @@
 // };
 
+// atomic still has a difference_type in the C++03 frozen headers
+// XFAIL: FROZEN-CXX03-HEADERS-FIXME
```
Thanks for helping with the CI.
Do you know why we test with C++03 in the first place?  `std::atomic` was introduced in C++11
